### PR TITLE
remote copy

### PR DESCRIPTION
**Pull request by vrothberg**
Implement `podman-remote cp` and break out the logic from the previously
added `pkg/copy` into it's basic building blocks and move them up into
the `ContainerEngine` interface and `cmd/podman`.

The `--pause` flag is now deprecated and turned into a nop.

Note that this commit is vendoring a non-release version of Buildah to
pull in updates to the copier package.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
One remaining bug that I will debug with Nalin.

File: cmd/podman/containers/cp.go
Comment (edsantiago) [inline cmd/podman/containers/cp.go:22]: It is no longer acceptable to double-space after periods. (It's a habit that has taken me a long time to break, but I accept the justification).

Comment (rhatdan) [reply cmd/podman/containers/cp.go:22]: I though this was just those of us that have actually used a typewriter.

Comment (vrothberg) [reply cmd/podman/containers/cp.go:22]: I am too old to change, @edsantiago

Comment (vrothberg) [reply cmd/podman/containers/cp.go:22]: > I thought this was just those of us that have actually used a typewriter.

I learned on a typewriter :)  I always use double spaces for text and now my brain parser is painfully dependent on it.

File: pkg/domain/infra/abi/containers_stat.go
Comment (rhatdan) [inline pkg/domain/infra/abi/containers_stat.go:118]: Why is this code here?

**Review by rhatdan — APPROVED** [review]
Impressive work
@containers/podman-maintainers PTAL

File: pkg/domain/infra/abi/containers_stat.go
Comment (vrothberg) [reply pkg/domain/infra/abi/containers_stat.go:118]: An old artifact, thanks for catching!

File: cmd/podman/containers/cp.go
Comment (vrothberg) [reply cmd/podman/containers/cp.go:335]: The terms can be confusing. We are copying "to" a container "from" a tar archive. So the error message is correct.

File: cmd/podman/containers/cp.go
Comment (Luap99) [inline cmd/podman/containers/cp.go:59]: Is there any reason why you did not use `flags.MarkDeprecated`

Comment (vrothberg) [reply cmd/podman/containers/cp.go:59]: I forgot its existence :)  If I need another push, I'll fix it :+1:

Comment (rhatdan) [reply cmd/podman/containers/cp.go:59]: Does MarkDeprecated hide the option?  I don't want users to see the option. Just want it to work so that there scripts would continue to work.

Comment (vrothberg) [reply cmd/podman/containers/cp.go:59]: Yes, that will hide the option but note that it's already hidden.

Comment (Luap99) [reply cmd/podman/containers/cp.go:59]: Yes it is hidden. If you still use it the message you specify with `MarkDeprecated` is emitted to stderr.
See `podman image diff --archive alpine > /dev/null` for example.

File: test/system/065-cp.bats
Comment (edsantiago) [inline test/system/065-cp.bats:101]: This is a useless test:  `/tmp/$dirname` has already been created (lines 91-92 above).

Worse, I find this command and comment confusing. What does it mean to "extract an archive to a file"? I would expect this to fail, and don't understand why it doesn't.

By adding the `exec` below I can see that podman seems to be stripping off `file.txt` and extracting into the parent directory. I think this is counterintuitive, and even dare say it is wrong behavior; but that's not my decision to make. If it is correct, or if it's impossible to fix, I think the test should be amended in some way (either with the `exec rm` or, preferably, with an extraction into a different directory) and, also, the comment should be expanded to clarify this weird behavior.

```diff
diff --git a/test/system/065-cp.bats b/test/system/065-cp.bats
index b1c3fbf8a..db9fbfff6 100644
--- a/test/system/065-cp.bats
+++ b/test/system/065-cp.bats
@@ -98,6 +98,7 @@ load helpers
     is "$output" "$rand_content"

     # Extract an archive to a file.
+    run_podman exec cpcontainer rm -rf /tmp/$dirname
     run_podman cp --extract=true $tar_file cpcontainer:/tmp/file.txt
     run_podman exec cpcontainer cat /tmp/$dirname/$rand_filename
     is "$output" "$rand_content"
```

**Review by edsantiago — COMMENTED** [review]
LGTM, nice work, but there's one broken test and very weird (unexpected) behavior.

File: cmd/podman/containers/cp.go
Comment (vrothberg) [reply cmd/podman/containers/cp.go:59]: Actually, using `MarkDeprecated(...)` is pretty ugly as it prints the warning twice for rootless podman (due to re-exec):
```
podman (run-950) $ ./bin/podman cp --pause=false Makefile test:/
Flag --pause has been deprecated, pause has been deprecated and and has no effect
Flag --pause has been deprecated, pause has been deprecated and and has no effect
```

**Comment by vrothberg** [comment]
Repushed.  Now both, --pause and --export, are deprecated and turned into nops.  I didn't use cobra's `MarkDeprecated` as it's printing the warning twice for rootless podman due to the reexec.

File: cmd/podman/containers/cp.go
Comment (TomSweeneyRedHat) [inline cmd/podman/containers/cp.go:103]: Maybe I have the exists() function from the new API stuck in my head, but I thought we'd similar functionality elsewhere in libpod?  I looked but didn't find it quickly.  This feels like a function that should live in util.

This is fine here for now, but something to ruminate on.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM
One question for consideration and unfortunately a lot of red in the tests.

File: cmd/podman/containers/cp.go
Comment (vrothberg) [reply cmd/podman/containers/cp.go:103]: Yes, there's a ContainerExists function.  I thought to put that into  a dedicated function (plus the checks) to consolidate the logic a bit further.

**Comment by vrothberg** [comment]
> Changes LGTM

Thanks for checking!

> One question for consideration and unfortunately a lot of red in the tests.

Should turn green now.  Removing the options made the help checks fail (as there are no options any more).

**Comment by rhatdan** [comment]
/lgtm
/hold
Go on vacation now...

**Comment by vrothberg** [comment]
> Go on vacation now...

One more day to go :)

**Comment by vrothberg** [comment]
One more time ... Ubuntu remote tests failed.  Rebased.

**Comment by vrothberg** [comment]
Note to self: close https://github.com/containers/podman/issues/4207 once this is merged

**Comment by vrothberg** [comment]
Tried to reproduce in the Google Cloud but it worked :hear_no_evil: Let's pray to the CI gods.

**Comment by vrothberg** [comment]
Oh man, that's frustrating.

**Comment by baude** [comment]
congrats @vrothberg

**Pull request merged**
